src/imex/transfer.rs

```diff
@@ -479,19 +492,72 @@ async fn transfer_from_provider(context: &Context, ticket: &Ticket) -> Result<()
     Ok(())
 }
 
+/// Run the get request
+async fn run_get_request(
+    context: &Context,
+    progress: &ProgressEmitter,
+    jobs: &Mutex<JoinSet<()>>,
+    ticket: Ticket,
+) -> anyhow::Result<Stats> {
+    let opts = ticket.as_get_options(Keypair::generate(), Some(default_derp_map()));
+    let request = AnyGetRequest::Get(GetRequest::all(ticket.hash()));
+    let connection = iroh::dial::dial(opts).await?;
+    let initial = fsm::start(connection, request);
+    use fsm::*;
+
+    let connected = initial.next().await?;
+    context.emit_event(ReceiveProgress::Connected.into());
+
+    // we assume that the request includes the entire collection
+    let (mut next, _root, collection) = {
+        let ConnectedNext::StartRoot(sc) = connected.next().await? else {
+            bail!("request did not include collection");
+        };
+
+        let (done, data) = sc.next().concatenate_into_vec().await?;
+        let data = Bytes::from(data);
+        let collection = Collection::from_bytes(&data)?;
+
+        context.emit_event(ReceiveProgress::CollectionReceived.into());
+        progress.set_total(collection.total_blobs_size());
+
+        (done.next(), data, collection)
+    };
+
+    // download all the children
+    let mut blobs = collection.blobs().iter();
+    let finishing = loop {
+        let start = match next {
+            EndBlobNext::MoreChildren(start) => start,
+            EndBlobNext::Closing(finishing) => break finishing,
+        };
+
+        // get the hash of the next blob, or finish if there are no more
+        let Some(blob) = blobs.next() else {
```

Comment:
this assumes that the request is complete without gaps. Which I guess is fine, but maybe there should be an assert for the assumption somewhere.

Reply:
how would you check for that?

Reply:
You could do an assert on the request. But then again you are creating the request a few lines earlier, so maybe just have a comment where the request is created? Otherwise check that child_index is consecutive.